core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
    * Get the level we can launch tasks according to delay scheduling, based on current wait time.
    */
   private def getAllowedLocalityLevel(curTime: Long): TaskLocality.TaskLocality = {
-    while (curTime - lastLaunchTime >= localityWaits(currentLocalityIndex) &&
-        currentLocalityIndex < myLocalityLevels.length - 1)
-    {
-      // Jump to the next locality level, and remove our waiting time for the current one since
-      // we don't want to count it again on the next one
-      lastLaunchTime += localityWaits(currentLocalityIndex)
-      currentLocalityIndex += 1
+    // Remove the scheduled or finished tasks lazily
+    def hasNotScheduledTasks(taskIndexes: ArrayBuffer[Int]): Boolean = {
+      var indexOffset = taskIndexes.size
+      while (indexOffset > 0) {
+        indexOffset -= 1
+        val index = taskIndexes(indexOffset)
+        if (copiesRunning(index) == 0 && !successful(index)) {
+          return true
+        } else {
+          taskIndexes.remove(indexOffset)
+        }
+      }
+      false
+    }
+    // It removes the empty lists after check
+    def hasMoreTasks(pendingTasks: HashMap[String, ArrayBuffer[Int]]): Boolean = {
+      val emptyKeys = new ArrayBuffer[String]
+      val hasTasks = pendingTasks.exists{
+        case (id: String, tasks: ArrayBuffer[Int]) =>
+          if (hasNotScheduledTasks(tasks)) {
+            true
+          } else {
+            emptyKeys += id
+            false
+          }
+      }
+      emptyKeys.foreach(x => pendingTasks.remove(x))
+      hasTasks
+    }
+
+    while (currentLocalityIndex < myLocalityLevels.length - 1) {
+      val moreTasks = myLocalityLevels(currentLocalityIndex) match {
+        case TaskLocality.PROCESS_LOCAL => hasMoreTasks(pendingTasksForExecutor)
+        case TaskLocality.NODE_LOCAL => hasMoreTasks(pendingTasksForHost)
+        case TaskLocality.NO_PREF => pendingTasksWithNoPrefs.isEmpty
+        case TaskLocality.RACK_LOCAL => hasMoreTasks(pendingTasksForRack)
+      }
+      if (!moreTasks) {
+        // Move to next locality level if there is no task for current level
+        lastLaunchTime = curTime
+        logDebug(s"No tasks for locality level ${myLocalityLevels(currentLocalityIndex)} " +
+          s"move to ${myLocalityLevels(currentLocalityIndex + 1)}")
+        currentLocalityIndex += 1
+      } else if (curTime - lastLaunchTime >= localityWaits(currentLocalityIndex)) {
+        // Jump to the next locality level, and remove our waiting time for the current one since
+        // we don't want to count it again on the next one
```

Comment:
I know this existed before your change, but can you change this comment to: "Jump to the next locality level, and reset lastLaunchTime so that the next locality wait timer doesn't immediately expire"? I think that would make this easier to understand